proposals/store/store.go

```diff
@@ -0,0 +1,178 @@
+package store
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+
+	"go.uber.org/zap"
+	"golang.org/x/exp/maps"
+
+	"github.com/spacemeshos/go-spacemesh/codec"
+	"github.com/spacemeshos/go-spacemesh/common/types"
+)
+
+var (
+	ErrNotFound       = errors.New("proposal not found")
+	ErrLayerEvicted   = errors.New("layer evicted")
+	ErrProposalExists = errors.New("proposal already exists")
+)
+
+type Store struct {
+	// number of layers to keep
+	capacity types.LayerID
+	logger   *zap.Logger
+
+	// guards access to data and evicted
+	mu      sync.RWMutex
+	evicted *types.LayerID
+	data    map[types.LayerID]map[types.ProposalID]*types.Proposal
+}
+
+type StoreOption func(*Store)
+
+func WithCapacity(capacity types.LayerID) StoreOption {
+	return func(s *Store) {
+		s.capacity = capacity
+	}
+}
+
+func WithEvictedLayer(layer types.LayerID) StoreOption {
+	return func(s *Store) {
+		s.evicted = &layer
+	}
+}
+
+func WithStoreLogger(logger *zap.Logger) StoreOption {
+	return func(s *Store) {
+		s.logger = logger
+	}
+}
+
+func New(opts ...StoreOption) *Store {
+	s := &Store{
+		data:     make(map[types.LayerID]map[types.ProposalID]*types.Proposal),
+		capacity: 2,
+		logger:   zap.NewNop(),
+	}
+	for _, opt := range opts {
+		opt(s)
+	}
+	var evicted uint32
+	if s.evicted != nil {
+		evicted = s.evicted.Uint32()
+	}
+	s.logger.Info("proposals store created", zap.Uint32("capacity", s.capacity.Uint32()), zap.Uint32("evicted", evicted))
+	return s
+}
+
+func (s *Store) IsEvicted(layer types.LayerID) bool {
+	return s.evicted != nil && *s.evicted >= layer
+}
+
+func (s *Store) OnLayer(layer types.LayerID) {
```

Comment:
could you please add metric for the number of proposals that we keep in this store, labeled by layer?